File: kirun-js/src/engine/function/system/date/AddSubtractTime.ts
import { DurationLikeObject } from 'luxon';
import { Schema } from '../../../json/schema/Schema';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { Parameter } from '../../../model/Parameter';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractDateFunction } from './AbstractDateFunction';
import { getDateTime } from './common';

export class AddSubtractTime extends AbstractDateFunction {
    public static readonly PARAMETER_YEARS_NAME = 'years';
    public static readonly PARAMETER_MONTHS_NAME = 'months';
    public static readonly PARAMETER_DAYS_NAME = 'days';
    public static readonly PARAMETER_HOURS_NAME = 'hours';
    public static readonly PARAMETER_MINUTES_NAME = 'minutes';
    public static readonly PARAMETER_SECONDS_NAME = 'seconds';
    public static readonly PARAMETER_MILLISECONDS_NAME = 'milliseconds';

    private readonly isAdd: boolean;

    constructor(isAdd: boolean) {
        super(
            isAdd ? 'AddTime' : 'SubtractTime',
            AbstractDateFunction.EVENT_TIMESTAMP,
            AbstractDateFunction.PARAMETER_TIMESTAMP,
            Parameter.of(
                AddSubtractTime.PARAMETER_YEARS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_YEARS_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_MONTHS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_MONTHS_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_DAYS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_DAYS_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_HOURS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_HOURS_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_MINUTES_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_MINUTES_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_SECONDS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_SECONDS_NAME).setDefaultValue(0),
            ),
            Parameter.of(
                AddSubtractTime.PARAMETER_MILLISECONDS_NAME,
                Schema.ofNumber(AddSubtractTime.PARAMETER_MILLISECONDS_NAME).setDefaultValue(0),
            ),
        );

        this.isAdd = isAdd;
    }
    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const timestamp = context
            .getArguments()
            ?.get(AbstractDateFunction.PARAMETER_TIMESTAMP_NAME);

        const dateTime = getDateTime(timestamp);
        const years = context.getArguments()?.get(AddSubtractTime.PARAMETER_YEARS_NAME);
        const months = context.getArguments()?.get(AddSubtractTime.PARAMETER_MONTHS_NAME);
        const days = context.getArguments()?.get(AddSubtractTime.PARAMETER_DAYS_NAME);
        const hours = context.getArguments()?.get(AddSubtractTime.PARAMETER_HOURS_NAME);
        const minutes = context.getArguments()?.get(AddSubtractTime.PARAMETER_MINUTES_NAME);
        const seconds = context.getArguments()?.get(AddSubtractTime.PARAMETER_SECONDS_NAME);
        const milliseconds = context
            .getArguments()
            ?.get(AddSubtractTime.PARAMETER_MILLISECONDS_NAME);

        const options = {
            years,
            months,
            days,
            hours,
            minutes,
            seconds,
            milliseconds,
        } as DurationLikeObject;

        let newDateTime;

        if (this.isAdd) {
            newDateTime = dateTime.plus(options);
        } else {
            newDateTime = dateTime.minus(options);
        }

        return new FunctionOutput([
            EventResult.outputOf(
                MapUtil.of(AbstractDateFunction.EVENT_TIMESTAMP_NAME, newDateTime.toISO()),
            ),
        ]);
    }
}

File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/date/AddSubtractTime.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.core.publisher.Mono;

public class AddSubtractTime extends AbstractDateFunction {

    public static final String PARAMETER_YEARS_NAME = "years";
    public static final String PARAMETER_MONTHS_NAME = "months";
    public static final String PARAMETER_DAYS_NAME = "days";
    public static final String PARAMETER_HOURS_NAME = "hours";
    public static final String PARAMETER_MINUTES_NAME = "minutes";
    public static final String PARAMETER_SECONDS_NAME = "seconds";
    public static final String PARAMETER_MILLISECONDS_NAME = "milliseconds";

    private final boolean isAdd;

    public AddSubtractTime(boolean isAdd) {
        super(isAdd ? "AddTime" : "SubtractTime",
                EVENT_TIMESTAMP,
                PARAMETER_TIMESTAMP,
                Parameter.of(PARAMETER_YEARS_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_YEARS_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_MONTHS_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_MONTHS_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_DAYS_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_DAYS_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_HOURS_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_HOURS_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_MINUTES_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_MINUTES_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_SECONDS_NAME,
                        Schema.ofNumber(AddSubtractTime.PARAMETER_SECONDS_NAME).setDefaultValue(new JsonPrimitive(0))),
                Parameter.of(PARAMETER_MILLISECONDS_NAME, Schema.ofNumber(AddSubtractTime.PARAMETER_MILLISECONDS_NAME)
                        .setDefaultValue(new JsonPrimitive(0))));

        this.isAdd = isAdd;
    }

    @Override
    protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters parameters) {

        String timestamp = parameters.getArguments()
                .get(AbstractDateFunction.PARAMETER_TIMESTAMP_NAME)
                .getAsString();

        ZonedDateTime dateTime = DateUtil.getDateTime(timestamp);

        Integer years = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_YEARS_NAME)
                .getAsInt();

        Integer months = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_MONTHS_NAME)
                .getAsInt();

        Integer days = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_DAYS_NAME)
                .getAsInt();

        Integer hours = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_HOURS_NAME)
                .getAsInt();

        Integer minutes = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_MINUTES_NAME)
                .getAsInt();

        Integer seconds = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_SECONDS_NAME)
                .getAsInt();

        Integer milliseconds = parameters.getArguments()
                .get(AddSubtractTime.PARAMETER_MILLISECONDS_NAME)
                .getAsInt();

        ZonedDateTime newDateTime;

        if (this.isAdd) {
            newDateTime = dateTime.plus(years, ChronoUnit.YEARS)
                    .plus(months, ChronoUnit.MONTHS)
                    .plus(days, ChronoUnit.DAYS)
                    .plus(hours, ChronoUnit.HOURS)
                    .plus(minutes, ChronoUnit.MINUTES)
                    .plus(seconds, ChronoUnit.SECONDS)
                    .plus(milliseconds, ChronoUnit.MILLIS);
        } else {
            newDateTime = dateTime.minus(years, ChronoUnit.YEARS)
                    .minus(months, ChronoUnit.MONTHS)
                    .minus(days, ChronoUnit.DAYS)
                    .minus(hours, ChronoUnit.HOURS)
                    .minus(minutes, ChronoUnit.MINUTES)
                    .minus(seconds, ChronoUnit.SECONDS)
                    .minus(milliseconds, ChronoUnit.MILLIS);
        }

        return Mono.just(new FunctionOutput(
                List.of(EventResult.outputOf(
                        Map.of(AbstractDateFunction.EVENT_TIMESTAMP_NAME,
                                new JsonPrimitive(
                                        newDateTime.format(DateUtil.ISO_DATE_TIME_FORMATTER)))))));
    }
}
